From 962d0500c4debaef43e5f146e47e08c66d851562 Mon Sep 17 00:00:00 2001
From: Michael Stahl <mstahl@redhat.com>
Date: Sat, 18 Feb 2012 00:06:33 +0100
Subject: [PATCH] fdo#39510: fix yet more layout crashes in ~SwRootFrm:

Call SwRootFrm::RemoveFtns with the proper boolean to also eliminate
EndNotes, and fix a faulty while loop in RemoveFtns that's been preventing
removal of footnotes in certain sections since the dawn of CVS history.
(regression from CWS swlayoutrefactoring, crashes on bugdoc from i#101776)
---
 sw/source/core/layout/ftnfrm.cxx |    2 +-
 sw/source/core/layout/newfrm.cxx |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/sw/source/core/layout/ftnfrm.cxx b/sw/source/core/layout/ftnfrm.cxx
index 7d3ea83..940bc11 100644
--- a/sw/source/core/layout/ftnfrm.cxx
+++ b/sw/source/core/layout/ftnfrm.cxx
@@ -1000,7 +1000,7 @@ void lcl_RemoveFtns( SwFtnBossFrm* pBoss, sal_Bool bPageOnly, sal_Bool bEndNotes
             if( pBody && pBody->Lower() )
             {
                 SwFrm* pLow = pBody->Lower();
-                while( pLow->GetNext() )
+                while (pLow)
                 {
                     if( pLow->IsSctFrm() && ( !pLow->GetNext() ||
                         ((SwSectionFrm*)pLow)->IsAnyNoteAtEnd() ) &&
diff --git a/sw/source/core/layout/newfrm.cxx b/sw/source/core/layout/newfrm.cxx
index 1e9b0e8..5f12ef5 100644
--- a/sw/source/core/layout/newfrm.cxx
+++ b/sw/source/core/layout/newfrm.cxx
@@ -610,7 +610,7 @@ SwRootFrm::~SwRootFrm()
     // also searches backwards to find the master of footnotes, they must be
     // considered to be owned by the SwRootFrm and also be destroyed here,
     // before tearing down the (now footnote free) rest of the layout.
-    AllRemoveFtns();
+    RemoveFtns(0, false, true);
 
     if(pBlink)
         pBlink->FrmDelete( this );
-- 
1.7.7.6

